### PR TITLE
Improve ChatEventHandler performance

### LINKED ISSUE #3980: Improve `DefaultChatEventHandler` performance by using `QueryMutableState::rawChannels`

### PR DESCRIPTION
**Pull request by adasiewiczr**
closes #3980 

### 🎯 Goal

Improve `DefaultChatEventHandler` performance.

### 🛠 Implementation details

The main improvement consists of passing `channels: Map<String, Channel>` instead of `channels: List<Channel>` to the event handler so we can do lookups in O(1).

1. Changed `ChatEventHandlerFactory::chatEventHandler` signature
2. Deprecated `QueryChannelsState.chatEventHandler` in favor of `QueryChannelsState.chatEventHandlerFactory`
3. Added additional `chatEventHandlerFactory` parameter to `ChatClient.queryChannelsAsState`
4. Simplified `CustomChatEventHandler` implementation

### 🧪 Testing
Test if sample apps work correctly, especially when adding/removing members and creating the channels.

#### General
- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [x] PR targets the `develop` branch
- [x] PR is linked to the GitHub issue it resolves

#### Code & documentation
- [x] Changelog is updated with client-facing changes
- [x] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [x] Affected documentation updated (KDocs, docusaurus, tutorial)

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [x] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### CODE REVIEW

File: stream-chat-android-state/src/main/java/io/getstream/chat/android/offline/plugin/state/querychannels/QueryChannelsState.kt [hunk 1]
```diff
@@ -51,6 +52,16 @@ public interface QueryChannelsState {
...
+    @Deprecated(
+        message = "Use chatEventHandlerFactory instead",
+        level = DeprecationLevel.WARNING,
+    )
     /** Instance of [ChatEventHandler] that handles logic of event handling for this [QueryChannelsState]. */
     public var chatEventHandler: ChatEventHandler?
+
+    /**
+     * [ChatEventHandler] factory, responsible for deciding whether the set of channels should be updated.
+     */
```
Comment (kanat) [inline stream-chat-android-state/src/main/java/io/getstream/chat/android/offline/plugin/state/querychannels/QueryChannelsState.kt:65]:
```suggestion
    /**
     * Factory that produces [ChatEventHandler], which is responsible for deciding whether the set of channels should be updated.
     */
```

Comment (kanat) [reply stream-chat-android-state/src/main/java/io/getstream/chat/android/offline/plugin/state/querychannels/QueryChannelsState.kt:65]: I guess you wanted to highlight the product, not a factory.

**Review by kanat — APPROVED** [review]
LGTM.

Didn't notice any changes in the sample apps' behavior.

**Review by MarinTolic — APPROVED** [review]
### ☑️Reviewer Checklist
- [x] UI Components sample runs & works
- [x] Compose sample runs & works
- [x] New feature tested and works
- [x] Release notes and docs clearly describe changes
- [x] All code we touched has new or updated KDocs

**Pull request merged**